### PR TITLE
[NEW MODULE] eCommerce - accept terms on payment page

### PR DESCRIPTION
**Pull request by mncoon**
@OSguard 
In addition to #1 here is the first part.

### CODE REVIEW

**Comment by mncoon** [comment]
Fail of https://travis-ci.org/OCA/l10n-germany/builds/40978897 is in checkout (payment site). 
I think this fail is ok, becaus checkout can't be done without activate the checkbox:

screenshot-192 168 222 213 8069 2014-11-14 10-58-37

**Comment by OSguard** [comment]
We have some customer with shops and we have no checkbox. We check this, you do not need a checkbox by law.
So there are two option to make the test green:
1) removing the checkbox because it is not needed
2) modify the test, that the test clicks on the checkbox. 

In the second case a new test to check if the failure raise if not check the checkbox is the way if we be total test driven. But i do not insist on this.

**Comment by mncoon** [comment]
@OSguard thank you for your reply. You are right, that the checkbox is not needed, so I removed it and  changed the text to "With click on the button "Pay now" I accept the terms and the revocation."

**Comment by mncoon** [comment]
@OSguard can you please have a look at Travis. I checkd the errors and I think it has nothing to do with the addon itself. Error happens after the order process on generating the pdf and sending the email:

```
2014-11-29 17:55:33,850 23658 INFO openerp_test openerp.addons.payment_transfer.models.payment_acquirer: Validated transfer payment for tx SO010: set as pending
2014-11-29 17:55:33,858 23658 INFO openerp_test werkzeug: 127.0.0.1 - - [29/Nov/2014 17:55:33] "POST /payment/transfer/feedback HTTP/1.1" 302 -
2014-11-29 17:55:34,387 23658 ERROR openerp_test openerp.addons.website.models.ir_http: 500 Internal Server Error:
```

`File "/home/travis/odoo-8.0/addons/report/models/report.py", line 269, in get_pdf
paperformat, specific_paperformat_args, save_in_attachment
File "/home/travis/odoo-8.0/openerp/api.py", line 237, in wrapper
return old_api(self, *args, **kwargs)
File "/home/travis/odoo-8.0/addons/report/models/report.py", line 442, in _run_wkhtmltopdf
wkhtmltopdf = [_get_wkhtmltopdf_bin()] + command_args + local_command_args
File "/home/travis/odoo-8.0/addons/report/models/report.py", line 54, in _get_wkhtmltopdf_bin
raise IOError
IOError`

It seems that something is wrong with getting wkhtmltopdf.

**Comment by OSguard** [comment]
@copmn please check here how to install wkhtml for test:

https://github.com/OCA/account-financial-reporting/blob/8.0/.travis.yml

just update the .travis.yml in your branch

**Comment by OSguard** [comment]
Can you please rebase of the changes to testing for #5

File: l10n_de_website_sale_accept_terms/views/views.xml [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+                                </div>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+            </t>
+        </template>
+
+        <!-- Template for imprint-page -->  
+        <template id="imprint" name="Imprint" page="True">
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/views/views.xml]: replace it with
<template id="website.imprint" name="Imprint" page="True">
then you can reach this page under "/page/imprint" and remove the "wsat_additions" controller

File: l10n_de_website_sale_accept_terms/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+import controllers
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/__init__.py]: from . import controllers

File: l10n_de_website_sale_accept_terms/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        """,
+    'author': 'copado MEDIA UG, Mathias Neef',
+    'website': 'http://www.copado.de',
+    'depends': ['website_sale'],
+    'data': [
+        'views/views.xml',
+    ],
+    'demo': [],
+    'installable': True,
+    'application': True,
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/__openerp__.py]: 'application': False,

File: l10n_de_website_sale_accept_terms/controllers/main.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        
+    @http.route(['/page/privacy', '/page/website.privacy'], type='http', auth='public', website=True)
+    def wsat_privacy(self, **kw):
+        return http.request.render('l10n_de_website_sale_accept_terms.privacy')
+
+    @http.route(['/page/imprint', '/page/website.imprint'], type='http', auth='public', website=True)
+    def wsat_imprint(self, **kw):
+        return http.request.render('l10n_de_website_sale_accept_terms.imprint')
+        
+wsat_additions()
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/controllers/main.py]: is not necessary and bad code style, please remove

File: l10n_de_website_sale_accept_terms/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp import http
+from openerp.http import request
+
+class wsat_additions(http.Controller):
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/controllers/main.py]: controller can be removed, se solution in template with id="webside.*"

File: l10n_de_website_sale_accept_terms/controllers/main.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    @http.route(['/page/imprint', '/page/website.imprint'], type='http', auth='public', website=True)
+    def wsat_imprint(self, **kw):
+        return http.request.render('l10n_de_website_sale_accept_terms.imprint')
+        
+wsat_additions()
+
+
+class wsat_popover(http.Controller):
+    @http.route('/popover/terms/', auth="public", type='http')
+    def wsat_popover_terms(self, **kw):
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/controllers/main.py]: is it possible to handle this in xml as well? i don't get the point why we need python code here

File: l10n_de_website_sale_accept_terms/i18n/de.po [hunk 7]
```diff
@@ -0,0 +1,179 @@
+# Translation of Odoo Server.
```
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/i18n/de.po:1]: l10n_de_website_sale_accept_terms/i18n/l10n_de_website_sale_accept_terms.pot as empty translation file is missing

**Comment by mncoon** [comment]
Now changes are done. I hope everything is fine. @OSguard thank you very much for your help with the module. I wish you a happy new year.

File: l10n_de_website_sale_accept_terms/i18n/de.po (hunk 7 above)
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/i18n/de.po:1]: not a empty file - add a empty translation file like:
https://github.com/OCA/l10n-germany/blob/7.0/l10n_de_country_states/i18n/l10n_de_country_states.po

File: l10n_de_website_sale_accept_terms/views/views.xml [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+        <!-- Add needed CSS and JS in template -->
+        <template id="assets_frontend" inherit_id="website.assets_frontend" name="Accept Terms on Checkout">
+          <xpath expr="." position="inside">
+                <link rel='stylesheet' href='/l10n_de_website_sale_accept_terms/static/src/css/style.css'/>
+                <script type="text/javascript" src="/l10n_de_website_sale_accept_terms/static/src/js/script.js"></script>
+          </xpath>
+        </template>
+
+        <!-- Replace the payment-site -->
```
Comment (vilder) [inline l10n_de_website_sale_accept_terms/views/views.xml:13]: I would recommend to just extend the payment site 

```
    <!-- Replace the payment-site -->
    <template id="wsga_payment" name="Checkout with Accept Terms" inherit_id="website_sale.payment">

         <xpath expr="//div[@class='row']" position="after">
                  <div class="wsga-mb30">
                      <h4 id="rev">Terms and revocation!</h4>
                      <div id="revlabel" class="form-group">
                          <p>With click on the button "Pay now" I accept the <span id="open-terms" class="wsga-cursor">terms</span> and the <span id="open-revo" class="wsga-cursor">revocation</span>.</p>
                      </div>
                  </div>
          <div id="wsga-terms" class="wsga-popover wsga-center" style="display:none;">
                <h3 class="popover-title">
                    Terms
                    <button id="close-terms" title="close" type="button" class="close">X</button>
                </h3>
                <iframe src="../../page/website.popover_terms"/>
            </div>


          <div id="wsga-revo" class="wsga-popover wsga-center" style="display:none;">
                <h3 class="popover-title">
                    Revocation
                    <button id="close-revo" title="close" type="button" class="close">X</button>
                </h3>
                <iframe src="../../page/website.popover_revocation/"/>
          </div>

        </xpath>
    </template>
```

Comment (mncoon) [inline l10n_de_website_sale_accept_terms/views/views.xml:13]: Hi @vilder you are right. 

Normaly extending is much better, but we have some restricts by law in germany that nothing must stand between the total amount and the payment button. That's the reason why I replaced it.

I hope this explaination is ok and if you have a better solution for this, you're welcome to tell.

**Comment by ntodorova** [comment]
Hi,
I install the module and try to see how its working, but I experienced a problem:
When I try to open the Terms page from the Footer I get the Revocation page and when I try to open the Revocation page I get err: 'This page does not exists...'. 
I think that this is caused, because line 204 and 226 from views.xml are equal.

And one more thing, I saw that at the Payment state I should see a drop down for choosing a delivery method, but I don`t see it. I configured a delivery method for my product, but do I have to make some more configurations in order to see it?
Greetings,
Niki

**Comment by mncoon** [comment]
@ntodorova thank you for your testing. I'd corrected a few things. 

To make the delivery method(s) visible you must install `website_sale_delivery`. After that you can configure as many delivery-methods as you want under `Warehouse -> Configuration -> Delivery Methods`. They should appear above "Choose your Payment Method".

**Comment by OSguard** [comment]
@ntodorova can please finish your review

**Comment by ntodorova** [comment]
:+1:

**Comment by OSguard** [comment]
i have the following remarks:

1) Translation: 'Durch klicken auf den Button "Jetzt bezahlen" akzeptiere ich die AGB und die Widerrufsbelehrung.'
but in German the button have the label "Jetzt kaufen" in my installation what is right by german law, so please change the translation in your module

2) please move module descriptio from **openerp** to README.rst

3) please add 'Odoo Community Association (OCA)' as author in **openerp**

**Comment by OSguard** [comment]
please use:
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
and look to runbot result

**Comment by mncoon** [comment]
@OSguard changes to readme.rst are done. But I can't reach the runbot result. It gives me an internal server error. Is there a way to restart?

File: .travis.yml [hunk 9]
```diff
@@ -19,11 +19,14 @@ virtualenv:
...
 install:
+  - git clone https://github.com/OCA/reporting-engine.git ${HOME}/reporting-engine -b ${VERSION}
   - git clone https://github.com/OCA/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools
+  - sudo wget http://sourceforge.net/projects/wkhtmltopdf/files/archive/0.12.1/wkhtmltox-0.12.1_linux-wheezy-amd64.deb -P /tmp/
+  - sudo dpkg -i /tmp/wkhtmltox-0.12.1_linux-wheezy-amd64.deb
   - export PATH=${HOME}/maintainer-quality-tools/travis:${PATH}
   - travis_install_nightly
-# example: dependency
-# - git clone https://github.com/OCA/webkit-tools -b ${VERSION} $HOME/webkit-tools
+-# example: dependency
```
Comment (OSguard) [inline .travis.yml]: your travis config is broken

Comment (OSguard) [inline .travis.yml]: remove the first char "-" in line 28 and 29

**Comment by mncoon** [comment]
@OSguard thank you! Now Travis looks good.

**Comment by OSguard** [comment]
Maybe @gurneyalex can tell us, what is wrong with the runbot

File: l10n_de_website_sale_accept_terms/views/views.xml (hunk 8 above)
Comment (OSguard) [inline l10n_de_website_sale_accept_terms/views/views.xml:13]: this is a good point, should we discuss this? Can you explain what to change?

Comment (mncoon) [inline l10n_de_website_sale_accept_terms/views/views.xml:13]: Yes, it is worth a further discussion.

This is what I changed:

Normally the whole acquirer-button-code (selectable payment methodes) is after the total amount. But this is not allowed in Germany.
To make it in consistency with our laws I seperated the code in two pieces:

https://github.com/copado/l10n-germany/blob/8.0/l10n_de_website_sale_accept_terms/views/views.xml#L34-L48

https://github.com/copado/l10n-germany/blob/8.0/l10n_de_website_sale_accept_terms/views/views.xml#L104-L111

And also a highlight around the products and prices:
https://github.com/copado/l10n-germany/blob/8.0/l10n_de_website_sale_accept_terms/views/views.xml#L57

With my previous knowledge about Odoo I don't find a way to do this in an extension.
But I can give a try a chance to do this with some xpath's.

**Comment by gurneyalex** [comment]
runbot is failing because of the following warning:

2015-06-23 09:09:22     WARNING     server  openerp.addons.base.module.module:784 check module l10n_de_website_sale_accept_terms: description is empty !

File: l10n_de_website_sale_accept_terms/readme.rst [hunk 10]
```diff
@@ -0,0 +1,83 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (gurneyalex) [inline l10n_de_website_sale_accept_terms/readme.rst:1]: The file needs to be renamed README.rst

this should fix the warning and make runbot green.

**Comment by legalsylvain** [comment]
@copmn : could you fix the minor problems ?

**Comment by pedrobaeza** [comment]
Is there still interest in this? If so, you need to rebase.

**Comment by mncoon** [comment]
At the moment there is no interest on this. Sorry.

**Comment by pedrobaeza** [comment]
OK, I close then.

**Pull request closed without merging**